### PR TITLE
[AutoDiff] Refactor ADContext into two classes.

### PR DESCRIPTION
**Pull request by bgogul**
This PR is part of the refactoring for https://bugs.swift.org/browse/TF-993. Specifically, this PR splits `ADContext` into two classes `ADContext` and `DifferentiationTransformer`. 
 - `ADContext` just keeps around the necessary state for the transformations.
 - `DifferentiationTransformer` performs the actual transformations.

This split is one of the necessary steps to split Differentiation.cpp into smaller manageable files.

### CODE REVIEW

**Comment by bgogul** [comment]
@swift-ci please test tensorflow

**Review by dan-zheng — COMMENTED** [review]
Main meta-comments: could you please share what header files you intend to create from `ADContext` and `DifferentiationTransformer`? I wonder if it's possible to reorganize `Differentiation.cpp` (TF-933) without adding a `DifferentiationTransformer` abstraction? If so, that sounds preferable to me!

The current interface of `ADContext` (lines [775](https://github.com/apple/swift/blob/bd7a725d83a15baf9e9db75e58f32f409ac4ccf7/lib/SILOptimizer/Mandatory/Differentiation.cpp#L775) to [1018](https://github.com/apple/swift/blob/bd7a725d83a15baf9e9db75e58f32f409ac4ccf7/lib/SILOptimizer/Mandatory/Differentiation.cpp#L1018)) doesn't seem to depend on `{JVP,VJP,Pullback}Emitter` or `LinearMapInfo` (dubbed "other stuff"), so it seems to me `ADContext` can be moved to its own header `ADContext.h`, and other stuff can be moved into their own header(s) importing `ADContext.h`.

---

Respectfully, I'm not sure the separation of `ADContext` and `DifferentiationTransformer` is the final desirable organization. Though I'm open to merging it to unblock TF-933, I don't think the initial organization needs to be perfect!

If we are to change code as part of TF-993, I think an alternative to creating `DifferentiationTransformer` is to move "canonicalization" methods to the existing [`Differentiation` class](https://github.com/apple/swift/blob/f7f6e9654e860bc6a0ff1f625dd6a62c3c32690d/lib/SILOptimizer/Mandatory/Differentiation.cpp#L7356), to limit the number of class abstractions. The [`LoadableByAddress` class](https://github.com/apple/swift/blob/bd7a725d83a15baf9e9db75e58f32f409ac4ccf7/lib/IRGen/LoadableByAddress.cpp#L1651) similarly defines helper functions.

A more significant change is TF-784: investigate subclassing `CanonicalizeInstruction` for canonicalizing `differentiable_function` instructions, to replace `ADContext::processDifferentiableFunctionInst`. This can allow us to remove auxiliary fields/functions like `ADContext::getDifferentiableFunctionInsts` and `{JVP,VJP}Emitter::visitDifferentiableFunctionInst`.

**Comment by bgogul** [comment]
> Main meta-comments: could you please share what header files you intend to create from `ADContext` and `DifferentiationTransformer`? I wonder if it's possible to reorganize `Differentiation.cpp` ([TF-933](https://bugs.swift.org/browse/TF-933)) without adding a `DifferentiationTransformer` abstraction? If so, that sounds preferable to me!

I expect the code in `DifferentialTransformer` to stay within `Differentiation.cpp`. I would just move the incarnation of `ADContext` in this PR to a separate header file. The split here makes `ADContext` a passive class that just stores some state.

> The current interface of `ADContext` (lines [775](https://github.com/apple/swift/blob/bd7a725d83a15baf9e9db75e58f32f409ac4ccf7/lib/SILOptimizer/Mandatory/Differentiation.cpp#L775) to [1018](https://github.com/apple/swift/blob/bd7a725d83a15baf9e9db75e58f32f409ac4ccf7/lib/SILOptimizer/Mandatory/Differentiation.cpp#L1018)) doesn't seem to depend on `{JVP,VJP,Pullback}Emitter` or `LinearMapInfo` (dubbed "other stuff"), so it seems to me `ADContext` can be moved to its own header `ADContext.h`, and other stuff can be moved into their own header(s) importing `ADContext.h`.

It is a bit more involved as there is cyclic dependency involving the derivating emitters and some functions like [canonicalizeDifferentiabilityWitness](https://github.com/apple/swift/blob/bd7a725d83a15baf9e9db75e58f32f409ac4ccf7/lib/SILOptimizer/Mandatory/Differentiation.cpp#L7224),  [emitDerivativeFunctionReference](https://github.com/apple/swift/blob/bd7a725d83a15baf9e9db75e58f32f409ac4ccf7/lib/SILOptimizer/Mandatory/Differentiation.cpp#L1768), etc. Splitting the `ADContext` class will make the subsequent refactoring easier. 

> Respectfully, I'm not sure the separation of `ADContext` and `DifferentiationTransformer` is the final desirable organization. Though I'm open to merging it to unblock [TF-933](https://bugs.swift.org/browse/TF-933), I don't think the initial organization needs to be perfect!

`ADContext` in this PR will move out, but the code in `DifferentiationTransformer` is still up in the air and may get reorganized more. Are you OK with that plan? 

> 
> If we are to change code as part of [TF-993](https://bugs.swift.org/browse/TF-993), I think an alternative to creating `DifferentiationTransformer` is to move "canonicalization" methods to the existing [`Differentiation` class](https://github.com/apple/swift/blob/f7f6e9654e860bc6a0ff1f625dd6a62c3c32690d/lib/SILOptimizer/Mandatory/Differentiation.cpp#L7356), to limit the number of class abstractions. The [`LoadableByAddress` class](https://github.com/apple/swift/blob/bd7a725d83a15baf9e9db75e58f32f409ac4ccf7/lib/IRGen/LoadableByAddress.cpp#L1651) similarly defines helper functions.

I am open to moving this to the `Differentiation` class. I wasn't sure if it is OK to have state in a derivative of `SILModuleTransform` class. That is why I created a `DifferentiationTransformer` class modeled after [PerformanceInlinerPass](https://github.com/apple/swift/blob/bd7a725d83a15baf9e9db75e58f32f409ac4ccf7/lib/SILOptimizer/Transforms/PerformanceInliner.cpp).

> A more significant change is [TF-784](https://bugs.swift.org/browse/TF-784): investigate subclassing `CanonicalizeInstruction` for canonicalizing `differentiable_function` instructions, to replace `ADContext::processDifferentiableFunctionInst`. This can allow us to remove auxiliary fields/functions like `ADContext::getDifferentiableFunctionInsts` and `{JVP,VJP}Emitter::visitDifferentiableFunctionInst`.

I think this PR will make it easier to deal with [TF-784](https://bugs.swift.org/browse/TF-784).

**Review by dan-zheng — APPROVED** [review]
LGTM, thanks for the clarifications!

@bgogul explained an advantage of not storing state in the `Differentiation` transform class is that `Differentiation::run` becomes more clearly idempotent.

File: lib/SILOptimizer/Mandatory/Differentiation.cpp [hunk 1]
```diff
@@ -926,14 +926,62 @@ class ADContext {
...
+  ADContext context;
+
   /// Promotes the given `differentiable_function` instruction to a valid
   /// `@differentiable` function-typed value.
   SILValue promoteToDifferentiableFunction(
       DifferentiableFunctionInst *inst, SILBuilder &builder, SILLocation loc,
       DifferentiationInvoker invoker);
 
 public:
+  /// Construct an DifferentiationTransformer for the given module.
```
Comment (rxwei) [inline lib/SILOptimizer/Mandatory/Differentiation.cpp]: We should add backquotes around references to any code.

File: lib/SILOptimizer/Mandatory/Differentiation.cpp [hunk 2]
```diff
@@ -1766,12 +1783,14 @@ reapplyFunctionConversion(
...
 emitDerivativeFunctionReference(
-    ADContext &context, SILBuilder &builder, SILAutoDiffIndices desiredIndices,
+    DifferentiationTransformer& transformer,
+    SILBuilder &builder, SILAutoDiffIndices desiredIndices,
     AutoDiffDerivativeFunctionKind kind, SILValue original,
     DifferentiationInvoker invoker,
     SmallVectorImpl<AllocStackInst *> &newBuffersToDealloc) {
 
   SILValue functionSource = original;
+  ADContext& context = transformer.getContext();
```
Comment (rxwei) [inline lib/SILOptimizer/Mandatory/Differentiation.cpp]: Our local convention is to put `&` on the variable name side.

File: lib/SILOptimizer/Mandatory/Differentiation.cpp [hunk 3]
```diff
@@ -1766,12 +1783,14 @@ reapplyFunctionConversion(
 /// FIXME: This is too complicated and needs to be rewritten.
 static Optional<std::pair<SILValue, SILAutoDiffIndices>>
 emitDerivativeFunctionReference(
-    ADContext &context, SILBuilder &builder, SILAutoDiffIndices desiredIndices,
+    DifferentiationTransformer& transformer,
```
Comment (rxwei) [inline lib/SILOptimizer/Mandatory/Differentiation.cpp]: Same here.

File: lib/SILOptimizer/Mandatory/Differentiation.cpp [hunk 4]
```diff
@@ -8020,8 +8039,9 @@ void Differentiation::run() {
   auto &astCtx = module.getASTContext();
   debugDump(module);
 
-  // A global differentiation context.
-  ADContext context(*this);
+  // A transformation helper.
+  DifferentiationTransformer transformer(*this);
+  ADContext& context = transformer.getContext();
```
Comment (rxwei) [inline lib/SILOptimizer/Mandatory/Differentiation.cpp]: `ADContext &context`

File: lib/SILOptimizer/Mandatory/Differentiation.cpp [hunk 5]
```diff
@@ -926,14 +926,62 @@ class ADContext {
...
       DifferentiationInvoker invoker);
 
 public:
+  /// Construct an DifferentiationTransformer for the given module.
+  explicit DifferentiationTransformer(SILModuleTransform &transform)
+    : transform(transform), context(transform) {}
+
+  ADContext& getContext() { return context; }
+
+  SILModuleTransform& getTransform() { return transform; }
```
Comment (dan-zheng) [inline lib/SILOptimizer/Mandatory/Differentiation.cpp]: I don't believe storing a `SILModuleTransform` is necessary as `ADContext` already does.
@bgogul: how about removing `DifferentiationTransformer:: transform` in a follow-up?

**Review by rxwei — COMMENTED** [review]
LGTM.

File: lib/SILOptimizer/Mandatory/Differentiation.cpp [hunk 6]
```diff
@@ -926,14 +926,62 @@ class ADContext {
...
   SILValue promoteToDifferentiableFunction(
       DifferentiableFunctionInst *inst, SILBuilder &builder, SILLocation loc,
       DifferentiationInvoker invoker);
 
 public:
+  /// Construct an DifferentiationTransformer for the given module.
+  explicit DifferentiationTransformer(SILModuleTransform &transform)
+    : transform(transform), context(transform) {}
+
+  ADContext& getContext() { return context; }
```
Comment (dan-zheng) [inline lib/SILOptimizer/Mandatory/Differentiation.cpp]: The current style is to put a space before `&`, not after. @bgogul: could you please address in this a follow-up?
```cpp
  ADContext &getContext() { return context; }
```

**Comment by bgogul** [comment]
Addressed the comments in https://github.com/apple/swift/pull/28568

**Pull request merged**